java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.ql

```diff
@@ -40,6 +40,18 @@ class UnsafeDeserializationConfig extends TaintTracking::Configuration {
       ma.getArgument(0) = pred.asExpr() and
       ma.getQualifier() = succ.asExpr()
     )
+    or
+    exists(CreateFromParcelMethod m, Variable v |
+      m.getEnclosingCallable().getDeclaringType() = v.getType() and
+      pred.asExpr() = v.getAnAssignedValue() and
```

Comment:
This seems odd -- can you explain what taint propagation you want from this?

As I understand it, if we have a

```
class ACreator implements Parcelable.Creator<A> {
  A createFromParcel(Parcel source) {
    ...
  }
}
```

and somewhere in the program we have

```
ACreator creator = getCreator();
```

This is conveying taint from `creator` to the parameter `source`? Why? Surely `source` is incoming data from the wire which may or may not be tainted, and `creator` having acquired some sort of taint of its own will taint the created instance, not the parcel?

Also `getAParameter()` is actually always exactly `getParameter(0)`, right?

Reply:
Hi @smowton,

<code>Parcelable.Creator</code> has a special usage pattern. As demonstrated in a [test program of this query](https://github.com/github/codeql/blob/42948c61259764a9ea196e8d2ea9b2515c5dbfd0/java/ql/test/query-tests/security/CWE-502/ParcelableEntity.java), its usage is like:
```
public class ParcelableEntity implements Parcelable {
   public static final Parcelable.Creator CREATOR = new Creator<ParcelableEntity>() {
        @Override
        public ParcelableEntity createFromParcel(Parcel parcel) {
             ...
   }
}
```

Then the class can be directly used by the invoking program:
```
   ParcelableEntity entity = (ParcelableEntity) getIntent().getParcelableExtra("jsonEntity");
```
<code>Creator</code> doesn't need to be initialized by the calling program and the Android framework does the wiring. 

Hope this answers your question.

Reply:
BTW, I've changed <code>getAParameter()</code> to <code>getParameter(0)</code>.

Reply:
So in this example, walk me through what taint edge you're adding?

Reply:
Draw me a complete taint graph where `entity` is involved? How is `entity` receiving taint which it then passes on to the parameter `parcel`? At what stage does an unsafe deserialization occur?

Reply:
Sorry I'm not sure how to draw a complete taint graph, and what I can think of is as follows:
<img width="1028" alt="Screen Shot 2021-07-13 at 1 10 01 PM" src="https://user-images.githubusercontent.com/37377780/125495950-7256df49-f9ec-420b-ab98-2eb865294035.png">

Please let me know if you need more information.

Reply:
That makes sense. I suggest you should implement this not as a dataflow step, which is counterintuitive because it draws a step *from* the `entity` to the parcel that it was created from, but instead as a source. If you see `getParcelableExtra(...)` cast to type `A` then that serves as indirect evidence that the `parcel` argument to `A$Creator` is a taint source.

Reply:
I've revamped the query of Android Gson detection as per your suggestion. I think the logic is much more clear now. Please review and let me know if it's not exactly what you want. Thanks.